pkg/agent/agent.go

```diff
@@ -63,47 +77,24 @@ func NewActionConfig(driver DriverType, token, namespace string) (*action.Config
 	config.BearerToken = token
 	config.BearerTokenFile = ""
 	clientset, err := kubernetes.NewForConfig(config)
-	store, err := createStorage(driver, namespace, clientset)
-	if err != nil {
-		return nil, err
-	}
+	store := storageForDriver(namespace, clientset)
 	actionConfig.RESTClientGetter = nil     // TODO replace nil with meaningful value
 	actionConfig.KubeClient = kube.New(nil) // TODO replace nil with meaningful value
 	actionConfig.Releases = store
 	actionConfig.Log = klog.Infof
 	return actionConfig, nil
 }
 
-func createStorage(driverType DriverType, namespace string, clientset *kubernetes.Clientset) (*storage.Storage, error) {
-	var store *storage.Storage
-	switch driverType {
-	case Secret:
-		d := driver.NewSecrets(clientset.CoreV1().Secrets(namespace))
-		d.Log = klog.Infof
-		store = storage.Init(d)
-	case ConfigMap:
-		d := driver.NewConfigMaps(clientset.CoreV1().ConfigMaps(namespace))
-		d.Log = klog.Infof
-		store = storage.Init(d)
-	case Memory:
-		d := driver.NewMemory()
-		store = storage.Init(d)
-	default:
-		return nil, fmt.Errorf("Invalid Helm drive type: %q", driverType)
-	}
-	return store, nil
-}
-
-func ParseDriverType(raw string) (DriverType, error) {
+func ParseDriverType(raw string) (StorageForDriver, error) {
 	switch raw {
 	case "secret", "secrets":
-		return Secret, nil
+		return StorageForSecrets, nil
 	case "configmap", "configmaps":
-		return ConfigMap, nil
+		return StorageForConfigMaps, nil
 	case "memory":
-		return Memory, nil
+		return StorageForMemory, nil
 	default:
-		return Memory, errors.New("Invalid Helm driver type: " + raw)
+		return StorageForMemory, errors.New("Invalid Helm driver type: " + raw)
```

Comment:
I feel like it might make more sense to return `nil` instead of `StorageForMemory` here, now that `ParseDriverType` returns a function instead of a quasi-enum/string. Can't say whether that should be part of this PR or a new one though.

Reply:
Agree, I'd return nil here (instead of a valid driver type)

Reply:
I had thought so to, but incorrectly also thought it was not possible. But you're right, the spec says as much too if I'd checked https://golang.org/ref/spec#Function_types . Updated the test to include checking for nil too. Thanks.